Date/Time Improvements

jon-stumpf opened the pull request:
I introduced a new function in the Parser class, nextDateTime(). This allowed me to eliminate duplicative code in many protocols. I also found some discrepancies in the comments for dates and time which I cleared up. I still have some anomalies that I am chasing and will raise issues to determine if these need to be fixed.

I merged the four and two digit year enum values to reduce the complexity of the function, as requested.  I find this to be a better compromise than using multiple flags as parameters.  I also removed millisecond information from three protocols as per discussion in Issue #2995.

I reverted the StarLink protocol changes to resolve the merge conflict.  This was a better option than what I did in my aborted pull request #2994.

tananaev commented on src/org/traccar/helper/Parser.java:
```diff
@@ -165,4 +167,122 @@ public double nextCoordinate() {
+        SMH_DMY,     // SSMMHHDDMMYYYY or SSMMHHDDMMYY
+
+        DMY_HMS,     // DDMMYYYYHHMMSS or DDMMYYHHMMSS
+        DMY_HMSms,   // DDMMYYYYHHMMSS.sss or DDMMYYHHMMSS.sss
```
It should all be uppercase. I suggest put double-S when there are milliseconds. Milliseconds are basically decimals of the seconds.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/helper/Parser.java:
```diff
@@ -165,4 +167,122 @@ public double nextCoordinate() {
+        YMD_HMS,     // YYYYMMDDHHMMSS or YYMMDDHHMMSS
+        YMD_HMSms,   // YYYYMMDDHHMMSS.sss or YYMMDDHHMMSS.sss
+
+        ISO8601      // YYYY-MM-DDTHH:MM:SS+HH:MM
```
This is never used. Please remove. I think for ISO parsing we can just use standard Java classes.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/helper/Parser.java:
```diff
@@ -165,4 +167,122 @@ public double nextCoordinate() {
+
+        switch (format) {
+            case HMS:
+                hour = nextInt(radix); minute = nextInt(radix); second = nextInt(radix); // (d{2})(d{2})(d{2})
```
Please remove comments. I think they are not very useful.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/helper/Parser.java:
```diff
@@ -165,4 +167,122 @@ public double nextCoordinate() {
+        }
+
+        if (year >= 0 && year < 100) {
+            year += 2000; // Future Y3K issue :)
```
Remove comment.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/protocol/AppelloProtocolDecoder.java:
```diff
@@ -36,8 +35,8 @@ public AppelloProtocolDecoder(AppelloProtocol protocol) {
-            .number("(dd)(dd)(dd)")              // date
-            .number("(dd)(dd)(dd).?d*,")         // time
+            .number("(dd)(dd)(dd)")              // date (yymmdd)
+            .number("(dd)(dd)(dd).?d*,")         // time (hhmmss.ms)
```
I think using "ms" is not correct here. Just remove milliseconds from format comment, because we are not capturing it anyway. Time format is rarely useful because the order is always the same anyway. Same for other similar cases.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/protocol/GpsmtaProtocolDecoder.java:
```diff
@@ -34,7 +34,7 @@ public GpsmtaProtocolDecoder(GpsmtaProtocol protocol) {
     private static final Pattern PATTERN = new PatternBuilder()
             .expression("([^ ]+) ")              // uid
-            .number("(d+) ")                     // time
+            .number("(d+) ")                     // time (unix time: seconds since 1970/01/01T00:00:00)
```
Unix time implies that it's from 1970... Please remove the explanation.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/protocol/Tk103ProtocolDecoder.java:
```diff
@@ -41,14 +41,14 @@ public Tk103ProtocolDecoder(Tk103Protocol protocol) {
             .expression(".{4},?")                // command
             .number("d*")                        // imei?
-            .number("(dd)(dd)(dd),?")            // date
+            .number("(dd)(dd)(dd),?")            // date (yymmdd or mmddyy?)
```
Are you sure it is MMDDYY and not DDMMYY? I think we need to clarify it and remove question mark.

jon-stumpf replied:
Done

tananaev commented on src/org/traccar/protocol/TotemProtocolDecoder.java:
```diff
@@ -41,20 +41,20 @@ public TotemProtocolDecoder(TotemProtocol protocol) {
             .number("(d+.d+)|")                  // vdop
             .number("(d+)|")                     // io status
-            .number("d+|")                       // time
+            .number("d+|")                       // battery time (units?)
```
Remove question. It doesn't add any value.

jon-stumpf replied:
Done

jon-stumpf commented:
I addressed all your comments.  I also was able to implement some further fixes.  In doing so, I chose to rebase some commits so that they would logically part of previous commits.

tananaev commented:
Everything is fine, except changes to the unit tests. Why did you change them?

Nearly all test cases are taken from actual devices. If it doesn't match documentation, it means that documentation is incorrect or not complete. Please never modify existing unit tests.

jon-stumpf commented:
I will revert all the changes to the test cases and make revisions to the protocols.

jon-stumpf commented:
Changes to protocol decoder tests have been reverted.

This is ready to merge.

tananaev commented:
Merged, thanks.
